python/mxnet/gluon/data/datareader.py

```diff
@@ -0,0 +1,33 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# coding: utf-8
+# pylint: disable=
+"""Dataset reader."""
+__all__ = ['DataReader']
+
+class DataReader(object):
```

Comment:
DataReader API shouldn't be like this. This is no different from Dataset.

I don't think datareader should provide __len__. It should just have a .next() method

Reply:
OK. At first I separated those classes by division of concern instead of interface. I agree that it would be better if the interfaces differ too.

Reply:
(e.g. SemEval2017 reader can read the dataset, while different dataset classes can treat them as datasets for different tasks, such as POS tagging and keyphrase extraction)

Reply:
Shall we make a data reader a generator? (i.e. `define __iter__`)

Reply:
It depends on what you want it to do.